lib-src/rpc2/rpc2-src/pack_helper.c

```diff
@@ -0,0 +1,383 @@
+/* BLURB lgpl
+
+                           Coda File System
+                              Release 5
+
+          Copyright (c) 1987-2016 Carnegie Mellon University
+                  Additional copyrights listed below
+
+This  code  is  distributed "AS IS" without warranty of any kind under
+the  terms of the  GNU  Library General Public Licence  Version 2,  as
+shown in the file LICENSE. The technical and financial contributors to
+Coda are listed in the file CREDITS.
+
+#*/
+
+
+#include <rpc2/pack_helper.h>
+#include <stdlib.h>
+#include <string.h>
+
+
+int unpack_int(BUFFER *buf, RPC2_Integer *ptr)
+{
+    if (buf->buffer + 4 > buf->eob)
+        return -1;
+    *ptr = ntohl(*(RPC2_Integer *)(buf->buffer));
+    buf->buffer += 4;
+    return 0;
+}
+
+
+int unpack_unsigned(BUFFER *buf, RPC2_Unsigned *ptr)
+{
+    if (buf->buffer + 4 > buf->eob)
+        return -1;
+    *ptr = ntohl(*(RPC2_Unsigned*)(buf->buffer));
+    buf->buffer += 4;
+    return 0;
+}
+
+
+int unpack_double(BUFFER *buf, RPC2_Double *ptr)
+{
+    if (buf->buffer + 8 > buf->eob)
+        return -1;
+    *ptr = *(RPC2_Double *)(buf->buffer);
+    buf->buffer += 8;
+    return 0;
+}
+
+
+int unpack_bound_bytes(BUFFER *buf, unsigned char *ptr, RPC2_Unsigned len)
+{
+    if (buf->buffer + len > buf->eob)
+        return -1;
+    memcpy(ptr, buf->buffer, len);
+    buf->buffer += _PAD(len);
+    return 0;
+}
+
+int unpack_unbound_bytes(BUFFER *buf, unsigned char *ptr)
+{
+    if (buf->buffer + 4 > buf->eob)
+        return -1;
+    *(RPC2_Byte *)ptr = *(RPC2_Byte *)(buf->buffer);
+    buf->buffer += 4;
+    return 0;
+}
+
+
+int unpack_string(BUFFER *buf, unsigned char **ptr, int who)
+{
+    if (buf->buffer + 4 > buf->eob)
+        return -1;
+
+    RPC2_Unsigned length = 1 + ntohl(*(RPC2_Integer *)(buf->buffer));
+    (buf->buffer) += 4;
+    if (buf->buffer + _PAD(length) > buf->eob)
+        return -1;
+    if (*(buf->buffer + length - 1) != '\0')
+        return -1;
+    /* If RPC2_String is the element of RPC2_Struct, mode should be NO_MODE. */
+	/* So mode should not be examined here. */
+	/* if (mode == IN_OUT_MODE && who == RP2_CLIENT) { */
+    if (who == STUBCLIENT) {
+		/* Just copy characters back */
+        memcpy(*ptr, buf->buffer, length);
```

Comment:
This is so dangerous, the caller doesn't know how long the received string is and probably has not allocated the proper amount of space. otoh it doesn't expect us to allocate either.

Maybe this should become, print a warning and return -1? nobody should really be using this code path to receive strings on the client.